### PR TITLE
70 sanction exceptions

### LINKED ISSUE #70: Sanction exceptions
There are a number of exceptions that preclude departures from sanctions in some cases. A way to enter and process these exceptions is needed.

Known exceptions:
- Joukkoliikennekohde (jore.jr_linja.linjlkohde) 9202 and 9204 should not be sanctioned for timing stops
- Lines 6918 and 6919 should not be included in anything at all. The procurement unit is 13/31K182NL

### PR DESCRIPTION
**Pull request by danieldunderfelt**
Implement sanction exceptions that make a departure not sanctionable by default. This is done as a pure database entity and no UI is necessary yet as the exceptions will probably not be updated often. Make a new task if these need to be user configurable in the future.

Closes #70 

FE PR: https://github.com/HSLdevcom/bultti-ui/pull/56

Changes:

- Add "transportSection" (joukkoliikennekohde, jr_linja.linjlkohde in JORE) to Departure and ObservedDeparture
- Add SanctionException entity
- Find matching SanctionException entities when collecting observed departure sanctions, set appliedSanctionAmount to 0 if a matching exception was found.
- Link matched SanctionException to Sanction

I added the exceptions to the database manually. Here they are as SQL:

```
insert into bultti.sanction_exception (id, exceptionAppliesToReason, exceptionValue, departureProperty)
values  ('2719c298-5a1a-4820-88bb-e4c58a4f323b', TIMING_STOP_VIOLATION, '9202', 'transportSection'),
        ('03f2ca96-46c5-4ebb-8f52-882bac651ac2', TIMING_STOP_VIOLATION, '9204', 'transportSection'),
        ('c435e9e7-3f24-4003-b393-752fca58f162', ALL, '6918', 'routeId'),
        ('f2462e9a-2cf1-4907-88f7-a1acd439bd3a', ALL, '6919', 'routeId');
```

Run the code first so that TypeORM creates the table, then insert the data.

### CODE REVIEW

**Comment by danieldunderfelt** [comment]
Suggestions for alternative names for "transportSection" are welcome. I didn't want to spent too much time on the name.

File: src/sanctioning/collectObservedDepartureSanctions.ts [hunk 1]
```diff
@@ -70,22 +71,37 @@ export async function collectObservedDepartureSanctions(
...
     let departureId = createDateJourneyKey(departure)
 
+    // Set applied amount to 0 if departure is excluded.
+    let sanctionException = getSanctionExceptionForDeparture(
+      departure,
+      sanctionReason,
+      sanctionExceptions
+    )
+
+    let appliedSanctionAmount = !!sanctionException ? 0 : sanctionAmount
```
Comment (Divinare) [inline src/sanctioning/collectObservedDepartureSanctions.ts:89]: I would do something like
```
Import isDepartureSanctionSkipped from 'sanctionExceptionUtils.ts'

let appliedSanctionAmount = isDepartureSanctionSkipped(deprature, sanctionReason
) ? 0 : sanctionAmount
```

/sanctionExceptionUtils.ts
```
const SANCTION_EXCEPTION_MAP = {
  all: ['6918', '6919'],
  timing_stop_violation: ['9202', '9204']
}

const isDepartureSanctionSkipped = (departure, sanctionReason) => {
    if (SANCTION_EXCEPTION_MAP['all']includes(departure.lineId) return true
    if (SANCTION_EXCEPTION_MAP['timing_stop_violation''].includes(departure.lineId) return true
    return false
}
```

This way exceptions would be nicely readable in code and ensured that are always applied. Now when exceptions are saved in DB, they need to be saved in dev/stage/prod/local db's

**Review by Divinare — COMMENTED** [review]
Code review done

File: src/sanctioning/collectObservedDepartureSanctions.ts (hunk 1 above)
Comment (danieldunderfelt) [reply src/sanctioning/collectObservedDepartureSanctions.ts:89]: Adding the `isDepartureSanctionSkipped` function is a very good idea, I thought I had already done it! :D

The reason for why these are in the database is that we may need to create a UI for this, to create and update them, if users want or if they are updated more frequently. Using a database entity also means that we can connect them to Sanction entities with relations. I don't see saving them in all environments as a problem, and in fact I'll add them to the seed function to ensure they exist.

**Review by Divinare — APPROVED** [review]
LGTM, didn't test though (will test it later)

**Pull request merged**